src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,942 @@
+                }
+                break;
+            }
+
+            // Remove the queue if it's emptied.
+            v.queue.is_empty()
+        });
+
+        result
+    }
+
+    /// Schedule delayed waking up on the specified key.
+    ///
+    /// Returns a future if it's needed to spawn a new async task to do the
+    /// delayed waking up. The caller should be responsible for executing
+    /// it.
+    fn handle_delayed_wake_up(
+        &self,
+        key_lock_wait_state: &mut KeyLockWaitState,
+        key: &Key,
+        wake_up_delay_duration_ms: u64,
+    ) -> Option<DelayedNotifyAllFuture> {
+        if let Some((_, start_time, delay_duration)) =
+            &mut key_lock_wait_state.delayed_notify_all_state
+        {
+            // There's already an async task spawned for handling delayed waking up on this
+            // key. Update its state to extend its delaying duration (until now
+            // + wake_up_delay_duration).
+            let new_delay_duration =
+                (start_time.saturating_elapsed().as_millis() as u64) + wake_up_delay_duration_ms;
+            delay_duration.store(new_delay_duration, Ordering::Release);
+            None
+        } else {
+            // It's needed to spawn a new async task for performing delayed waking up on
+            // this key. Return a future to let the caller execute it in a
+            // proper thread pool.
+            let notify_id = self.allocate_internal_id();
+            let start_time = Instant::now();
+            let delay_ms = Arc::new(AtomicU64::new(wake_up_delay_duration_ms));
+
+            key_lock_wait_state.delayed_notify_all_state =
+                Some((notify_id, start_time, delay_ms.clone()));
+            Some(Box::pin(self.clone().async_delayed_notify_all(
+                key.clone(),
+                start_time,
+                delay_ms,
+                notify_id,
+            )))
+        }
+    }
+
+    fn allocate_internal_id(&self) -> u64 {
+        self.inner.id_allocated.fetch_add(1, Ordering::SeqCst)
+    }
+
+    async fn async_delayed_notify_all(
+        self,
+        key: Key,
+        start_time: Instant,
+        delay_ms: Arc<AtomicU64>,
+        notify_id: u64,
+    ) -> Option<Box<LockWaitEntry>> {
+        // tokio::task::yield_now().await;
+        let mut prev_delay_ms = 0;
+        // The delay duration may be extended by later waking-up events, by updating the
+        // value of `delay_ms`. So we loop until we find that the elapsed
+        // duration is larger than `delay_ms`.
+        loop {
+            let current_delay_ms = delay_ms.load(Ordering::Acquire);
+            if current_delay_ms == 0 {
+                // Cancelled.
+                return None;
+            }
+
+            if current_delay_ms <= prev_delay_ms
+                || (start_time.saturating_elapsed().as_millis() as u64) >= current_delay_ms
+            {
+                // Timed out.
```

Comment:
Here it checks whether `wake-up-delay-duration` has been past since the most recent waking up on the key, so that it can perform a delayed_notify_all. Is `wait-for-lock-timeout` useful here?

Reply:
If the `delay_ms` is updated continuously, and the wait time has exceeded the `wait-for-lock-timeout` which is the maximum wait time, would it be possible that this loop is still being executed though the waiter has been canceled?

Reply:
Yes. But it's still possible that there are some new waiters added later and need to be woken up after delay. So in this case we still continue. For example, consider the queue:
[10, 20, 30]
Wake up: [20*, 30*] and schedule `delayed_notify_all` (item marked with * means need to be woken up after delay)
Key is locked by other transaction and some new waiters are added: [20*, 30*, 40, 50]
Wake up: [30*, 40*, 50*]
Wait for 30 timeout: [(30 stale), 40*, 50*]
And then call `delayed_notify_all`.

In this case, though 30 (which exists when we schedule `delayed_notify_all`) is expired, but it still need to cancel 40 and 50 which is added later.